File: webapp/webapi/models/article.py
from datetime import datetime

from sqlalchemy import (
    Column,
    String,
    DateTime,
    ForeignKey
)

from sqlalchemy.orm import (
    relationship, backref
)

from marshmallow_sqlalchemy import auto_field

from ..configs import db, ma

from .category import Category
from .author import Author


class Article(db.Model):
    __bind_key__ = "articles"
    __tablename__ = "articles"

    id = Column(String, primary_key=True)

    #Constraint must be at least 1 character and at most 250
    title = Column(String)

    #Constraint must be at least 50 character and at most 500
    description = Column(String) #nullable=True

    tags = Column(String)
    #Constraint must be at least 100 characters

    content = Column(String)
    #add constraint here available state are: active, fired, unactive 

    status = Column(String, default='active') 

    create_at = Column(
                    DateTime, 
                    default=datetime.utcnow, 
                    onupdate=datetime.utcnow
    )

    publish_at = Column(
                    DateTime, 
                    nullable = True
                )

    update_at = Column(
                    DateTime,
                    nullable = True 
                )
    
    
    # categories = relationship(
    #     "Category", 
    #     backref=backref("articles"),
    #     # cascade="all, delete, delete-orphan",
    #     # single_parent=True,
    #     order_by="desc(Category.name)"
    # )

    # authors = relationship(
    #     "Author", 
    #     backref=backref("authors"),
    #     # cascade="all, delete, delete-orphan",
    #     # single_parent=True,
    #     order_by="desc(Author.surname)"
    # )

    

    def __repr__(self) -> str:
        return """Article(
            id={}, title='{}', description='{}',
            tags={}, content='{}', status={}, create_at={},
            publish_at={}, update_at={}
            )""".format(
            self.id, self.title, self.description, 
            self.tags, self.content, self.status,
            self.create_at, self.publish_at, self.update_at
        )


class ArticleSchema(ma.SQLAlchemySchema):
    class Meta:
        model = Article
        load_instance = True #Optional: deserialize to model instance
    id = auto_field()
    title = auto_field()
    description = auto_field()
    tags = auto_field()
    content = auto_field()
    status = auto_field()
    create_at = auto_field()
    publish_at = auto_field()
    update_at = auto_field()
    # categories = auto_field()
    # authors = auto_field()
 

File: webapp/frontend/views/static/js/types/article.js
import Author from "./author.js";
import ArticleCategory from "./category.js";
import ArticleReference from "./reference.js";
import ArticleTag from "./tag.js";
class Article {
    constructor(id, title, description, tags, content, status, createAt, publishAt, updateAt, coverImage, categories, authors, references) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.content = content;
        this.status = status;
        this.coverImage = coverImage;
        this.createAt = new Date(Date.parse(createAt));
        this.publishAt = new Date(Date.parse(publishAt));
        this.updateAt = new Date(Date.parse(updateAt));
        for (var tag of tags) {
            this.tags.push(new ArticleTag(tag));
        }
        for (var category of categories) {
            this.categories.push((new ArticleCategory()).fromJson(category));
        }
        for (var author of authors) {
            this.authors.push((new Author()).fromJson(author));
        }
        for (var reference of references) {
            this.references.push((new ArticleReference()).fromJson(reference));
        }
    }
    fromJson(article) {
        this.id = article['id'];
        this.title = article['title'];
        this.description = article['description'];
        this.content = article['content'];
        this.status = article['status'];
        this.coverImage = article['cover_image'];
        this.createAt = new Date(Date.parse(article['create_at']));
        this.publishAt = new Date(Date.parse(article['publish_at']));
        this.updateAt = new Date(Date.parse(article['update_at']));
        for (var tag of article['tags']) {
            this.tags.push(new ArticleTag(tag));
        }
        for (var category of article['categories']) {
            this.categories.push((new ArticleCategory()).fromJson(category));
        }
        for (var author of article['authors']) {
            this.authors.push((new Author()).fromJson(author));
        }
        for (var reference of article['references']) {
            this.references.push((new ArticleReference()).fromJson(reference));
        }
    }
    toJson() {
        return {
            'id': this.id,
            'title': this.title,
            'description': this.description,
            'tags': this.tags.map((tag) => tag.tagName),
            'content': this.content,
            'status': this.status,
            'cover_image': this.coverImage,
            'create_at': this.createAt.toISOString(),
            'publish_at': this.publishAt.toISOString(),
            'update_at': this.updateAt.toISOString(),
            'categories': this.categories.map((category) => category.toJson()),
            'authors': this.authors.map((author) => author.toJson()),
            'references': this.references.map((reference) => reference.toJson())
        };
    }
}
export default Article;
